Adjust datastore refresh behaviour to reduce deadlock risk

ThrawnCA opened the pull request:
- Detect when resource structure is unchanged and run a TRUNCATE with a short lock timeout, instead of dropping and recreating the table.
- Also add Tabulator guessed types to the data dictionary so they will be preserved.

ThrawnCA committed:
adjust datastore refresh behaviour to reduce deadlock risk

- Detect when resource structure is unchanged and run a TRUNCATE with a short lock timeout,
instead of dropping and recreating the table.
- Also add Tabulator guessed types to the data dictionary so they will be preserved.

wardi reviewed (approved):
Looks like a great change, we may have hit the deadlocks mentioned in this PR when we needed to clear out the queue

ThrawnCA committed:
[QOLSVC-3902] pre-emptively lock resource for editing when updating metadata

- When retrieving the resource extras, set the 'FOR UPDATE' flag so we get a write-lock immediately.
This avoids the risk of deadlocking if two processes try to update the same resource at the same time.

ThrawnCA commented:
@wardi I have added a change to pre-emptively acquire write locks when retrieving resources to update their extras.

I think this was overlooked when implementing https://github.com/ckan/ckan/issues/3245

wardi commented:
@ThrawnCA package_revise etc. take a lock on the package when making changes, including changes to resources. IIUC it would be safer to lock the package here instead.

ThrawnCA commented:
> IIUC it would be safer to lock the package here instead.

But we're not editing the package here. We only need a resource lock in order to finish our piece of work, at which point we can release the lock and let any other process go on its way.

- Scenario one, we're first. We get a resource write lock, another process grabs the package lock and maybe some other resource locks, then has to wait for us. We complete the update, release the lock, then the other process picks it up and completes its update.
- Scenario two, we're second. Another process grabs the process lock and resource lock. We can't get the lock we want, so we wait until it completes, then we get the resource write lock and do our update.

There's no risk of deadlock either way, and the package lock doesn't matter to us.

Without the change, though, we're only requesting the resource _read_ lock at first, so there is a chance for both processes to successfully acquire that lock and then both wait for the write lock (but neither can proceed, so the read locks aren't released until the deadlock is detected and broken).

wardi commented:
Yes it sounds like your approach will avoid the deadlock you describe, which is the most important fix.

I'm thinking of the case where we lose an update:
1. e.g. package_revise or resource_patch are used to update a resource field so they use the `{'for_update': True}` context parameter to `package_show`, acquiring a package lock and returning all the package+resource metadata to the python code for processing
2. this code updates the same resource, ignoring the package lock
3. package_revise or resource_patch modify and send their copy of the resource metadata to `package_update`, reverting the change in step 2

If we take the package lock we could avoid this problem the small cost of possibly slowing down some concurrent updates, right?

ThrawnCA commented:
>     1. e.g. package_revise or resource_patch are used to update a resource field so they use the `{'for_update': True}` context parameter to `package_show`, acquiring a package lock and returning all the package+resource metadata to the python code for processing
> 
>     2. this code updates the same resource, ignoring the package lock

But if step 1 is retrieving resource data, then it must be acquiring the resource locks as well as the package lock.

So when the XLoader code tries to acquire a resource _write_ lock, it will have to first wait until the package update completes.

wardi commented:
Yes that's another approach, in core we could lock all the tables in `package_show` when `{'for_update': True}` is in the context. That's now how it's implemented right now.

ThrawnCA commented:
> Yes that's another approach, in core we could lock all the tables in `package_show` when `{'for_update': True}` is in the context. That's now how it's implemented right now.

Core changes shouldn't be needed, I think. If a package update is retrieving resource data along with the package, then it must be acquiring at least a resource read lock. Which will mean the XLoader, which wants a write lock thanks to this change, will have to wait.

wardi commented on ckanext/xloader/loader.py:
```diff
@@ -79,6 +79,49 @@ def detect_encoding(file_path):
+    '''
+    # drop the generated '_id' field
+    for index in range(len(existing_fields)):
+        if existing_fields[index]['id'] == '_id':
+            existing_fields.pop(index)
```
removing an item from the list then indexing based on the old list length can cause an IndexError, right?

ThrawnCA replied:
We immediately `break` after popping it.

wardi replied:
yeah caught that just after posting my comment :-)
